### ISSUE TITLE
activationEvents.workspaceContains doesn't fire for directory

### ISSUE THREAD

**Issue by flynnwt** [Microsoft/vscode#2739]
Is this expected?  Also, a general way to detect all workspace-open events, like 'workspaceContains:*' or an onDidChangeWorkspace event would be useful.

**Comment by egamma** [comment]
Yes currently this event is only triggered for files. The Scenario is when there is a file like `.editorconfig`.
What is your scenario, where you want to be activated for a particular directory?

CC @alexandrudima

**Comment by flynnwt** [comment]
I initially was trying to use the fact that a directory exists to enable an extension (for backing up timestamped edits), and just using the user settings for configuration.  I did add an empty file for now to trigger it.  If I wanted to make the directory configurable, I would need a separate config file anyway, or a notification on any workspace opening, to check dynamically.

**Comment by egamma** [comment]
For this scenario I would tolerate using this activation event: https://code.visualstudio.com/docs/extensionAPI/activation-events#_activationevents

**Comment by flynnwt** [comment]
Thanks, I did play with that but didn't realize from the description that it occurs on every workspace open, not just initial startup.  So the '*'  activation does allow any kind of per-workspace checking to be done before enabling an extension.

**Comment by iambibhas** [comment]
@egamma in my case, I want the extension to activate only when current directory is a git repository. i.e. it'll have the `.git/` directory in current directory. I've tried putting `"workspaceContains:.git/config"` but that does nothing and the commands are activated anyway. I don't think I understand how this `activationEvents` works. :|

**Comment by egamma** [comment]
reopening

**Comment by alexandrudima** [comment]
@iambibhas `workspaceContains:X` gets fired for all X where X is a top level file. If you open a folder in VSCode that contains:
* .git (folder)
* bla.txt (file)
* foo (file)

the following events will be fired when the folder is opened: `workspaceContains:bla.txt` and `workspaceContains:foo`

**Comment by iambibhas** [comment]
@alexandrudima so then shouldn't the activation fail if `workspaceContains:.git` is present? Why does it still get activated?

**Comment by alexandrudima** [comment]
@iambibhas That is the actual feature request, that we fire activation events for folders too.

The code today only does so for files:
https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/node/extensionHostMain.ts#L249
https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/pfs.ts#L188
https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/pfs.ts#L168

**Comment by tboby** [comment]
I'd like to extend this request to not only trigger for folders inside the workspace, but for the full path to the folders inside the workspace. This would align with how languages can use `filenamePatterns` to reference folders above their workspace.

E.g.

```
"filenamePatterns": [
     "**/stellaris/**/*.txt"
]
```
Allows me to activate when you open a file `/stellaris/myproject/script.txt` (in combination with `onLanguage`

I'd ideally like to be able to activate my extension when you open a folder `/stellaris/myproject/`

Not sure if this should be a new request for an activation event `"workspacePattern":"**/stellaris/**"` instead however.
